repository/src/fetch.rs

```diff
@@ -0,0 +1,269 @@
+                last_finalization_proof = proof.clone();
+                last_finalized_commit_hash = *block_commit_hashes.last().unwrap();
+                hit = true;
+            }
+        }
+
+        // If failed, finalize the last-last block.
+        if !hit && block_commit_hashes.len() > 2 {
+            let last_block = csv.get_block_headers().last().unwrap().clone();
+            last_finalization_proof = LastFinalizationProof {
+                proof: last_block.prev_block_finalization_proof.clone(),
+                height: last_block.height - 1,
+            };
+            last_finalized_commit_hash = block_commit_hashes[block_commit_hashes.len() - 2];
+        }
+    }
+
+    // step 2: create `a` and `b` branches.
+    for (remote_name, branch_name, commit_hash, commit) in next_step_branches {
+        let branch_displayed = format!(" {}/{}(at {})", remote_name, branch_name, commit_hash);
+        match &commit {
+            Commit::Agenda(agenda) => {
+                if agenda.height == last_finalization_proof.height + 1 {
+                    let branch_name = format!(
+                        "a-{:?}",
+                        commit
+                            .to_hash256()
+                            .to_string()
+                            .truncate(COMMIT_TITLE_HASH_DIGITS)
+                    );
+                    this.raw.create_branch(branch_name, commit_hash).await?;
+                }
+            }
+            Commit::AgendaProof(agenda_proof) => {
+                if agenda_proof.height == last_finalization_proof.height + 1 {
+                    let branch_name = format!(
+                        "a-{:?}",
+                        commit
+                            .to_hash256()
+                            .to_string()
+                            .truncate(COMMIT_TITLE_HASH_DIGITS)
+                    );
+                    this.raw.create_branch(branch_name, commit_hash).await?;
+                }
+            }
+            Commit::Block(block_header) => {
+                if block_header.height == last_finalization_proof.height + 1 {
+                    let branch_name = format!(
+                        "b-{:?}",
+                        commit
+                            .to_hash256()
+                            .to_string()
+                            .truncate(COMMIT_TITLE_HASH_DIGITS)
+                    );
+                    this.raw.create_branch(branch_name, commit_hash).await?;
+                }
+            }
+            x => warn!("incomplete remote branch {}: got {:?}", branch_displayed, x),
+        }
+    }
+
+    // step 3: update finalization
+    this.raw.checkout_clean().await?;
+    this.raw
+        .move_branch(FINALIZED_BRANCH_NAME.into(), last_finalized_commit_hash)
+        .await?;
+    this.raw
+        .move_branch(FP_BRANCH_NAME.into(), last_finalized_commit_hash)
+        .await?;
+    this.raw.checkout(FP_BRANCH_NAME.into()).await?;
```

Comment:
We checkout to FP_BRANCH and finished this method.
I think we have to finish this method with our current working branch. 
-> I think we should re-checkout to the working branch.(like real-git)

Reply:
There is no 'working branch'. Where should I checkout then? Note that the `work` branch belongs to the clone of this repository. (user doesn't directly checkout and add commits to the `work` branch, but just push it)